Implement backend for Help section in Sensei Home.

aaronfc opened the pull request:
Fixes #5742 

### Changes proposed in this Pull Request
* Implement backend for Help section.
* Added `sensei_home_support_ticket_creation_upsell_show` filter.

### Testing instructions
* Run `curl -sX GET --user aaron:development --insecure 'https://test2.local/wp-json/sensei-internal/v1/home' | jq .help` and confirm that output is as expected. ([Check how to test curl in local more easily](https://evergreenp2.wordpress.com/2022/03/03/testing-cheatsheet/#test-rest-api-in-local-easily))
* Disbale `sensei_home_support_ticket_creation_upsell_show` with a snippet.
* Run the same curl from above and confirm that output changed correctly.

### New hooks/filters
- Added `sensei_home_support_ticket_creation_upsell_show` filter to allow to disable upsell action to Sensei Pro.

### Screenshot / Video
**Default:**
```
$ curl -sX GET --user aaron:development --insecure 'https://test2.local/wp-json/sensei-internal/v1/home' | jq .help
[
  {
    "title": "Get the most out of Sensei",
    "items": [
      {
        "title": "Sensei Documentation",
        "url": "https://senseilms.com/documentation/",
        "icon": null,
        "extra_link": null
      },
      {
        "title": "Support forums",
        "url": "https://wordpress.org/support/plugin/sensei-lms/",
        "icon": null,
        "extra_link": null
      },
      {
        "title": "Create a support ticket",
        "url": null,
        "icon": "lock",
        "extra_link": {
          "label": "Upgrade to Sensei Pro",
          "url": "https://senseilms.com/pricing/"
        }
      }
    ]
  }
]
```

**With `sensei_home_support_ticket_creation_upsell_show` disabled:**
```
$ curl -sX GET --user aaron:development --insecure 'https://test2.local/wp-json/sensei-internal/v1/home' | jq .help
[
  {
    "title": "Get the most out of Sensei",
    "items": [
      {
        "title": "Sensei Documentation",
        "url": "https://senseilms.com/documentation/",
        "icon": null,
        "extra_link": null
      },
      {
        "title": "Support forums",
        "url": "https://wordpress.org/support/plugin/sensei-lms/",
        "icon": null,
        "extra_link": null
      },
      {
        "title": "Create a support ticket",
        "url": "https://senseilms.com/contact/",
        "icon": "lock",
        "extra_link": null
      }
    ]
  }
]
```

aaronfc committed:
Add Sensei Home Help domain logic.

aaronfc committed:
Integrate Help section with controller + tests.

aaronfc committed:
Replace filter with class existance check for Sensei Pro detection.

- Basing check on filter would still need a fallback to class check in
  case an old version of Sensei Pro is being used.
- Positive case (when `\Sensei_Pro\Sensei_Pro` exists) can't be tested
  with unit tests. Will need manual test.

fjorgemota commented on includes/admin/home/help/class-sensei-home-help-provider.php:
```diff
@@ -0,0 +1,65 @@
+		} else {
+			$extra_link = new Sensei_Home_Help_Extra_Link( __( 'Upgrade to Sensei Pro', 'sensei-lms' ), 'https://senseilms.com/pricing/' );
+		}
+		return new Sensei_Home_Help_Item( __( 'Create a support ticket', 'sensei-lms' ), $url, 'lock', $extra_link );
```
If the user has Sensei Pro we don't want to have the lock icon on the support ticket link, right?

aaronfc replied:
Nice catch! Didn't thought about that association but yeah, lock icon only makes sense when ticket creation is "locked" due not having Sensei Pro.

Changed here: 4e3449805

fjorgemota reviewed (commented):
Hey @aaronfc, thanks for looking into this. I found one small possible logic mistake, and I'd like to comment on this:

> Sensei Pro detection is based on the existance of \Sensei_Pro\Sensei_Pro class. We can discuss on alternatives for this, but using a filter still would need some class check for the case when an old version of Sensei Pro is being used.

I don't think we should particularly bother about supporting an old version of Sensei Pro here. We already don't do that, anyway, and the idea is to always recommend users to have the very latest versions of Sensei Pro and Sensei LMS together, in case of issues. 

I also confirmed with Ronnie and this line of thinking is okay: p1664322416527539-slack-C8QJAM52B

I also don't think we need to do class checks here, either.

The standard approach we use  is to just call `do_action`/`apply_filters` on Sensei LMS, and on Sensei Pro run `add_action`/`add_filter` to add a hook changing the behavior of the code on Sensei LMS. We already did that many times in the past, and even in this project we're doing the same thing again already, see: 1803-gh-Automattic/sensei-pro - We had some PRs similar to that in the past projects we worked on, like in the Interactive Video Blocks, for instance.

This approach also has the advantage of only changing things when Sensei Pro is activated, which is great. I'm not sure if `class_exists` would work well in this scenario, to be honest. The possibility to decouple the code between the two plugins is the biggest advantage, though.

That's why we don't need anything like a `Sensei_Pro_Detector`, at all. In practice, the less the code knows about Sensei Pro, the better, especially because any test focusing on detecting that easily breaks in a dev environment like the one I have thanks to the fact that Sensei Pro and Sensei Core is both always installed, and, well, I like running tests in my own machine. :D 

Does it make sense?

--- 

With that said, the rest of the code looks generally okay. I admit that I didn't take a very deep look, though, so I'll test/review more after we have the Sensei Pro detection thing more...aligned. =)

aaronfc committed:
Use lock icon only when sensei pro is not installed.

aaronfc commented:
Hi @fjorgemota thanks for your review!

> The standard approach we use is to just call do_action/apply_filters on Sensei LMS, and on Sensei Pro run add_action/add_filter to add a hook changing the behavior of the code on Sensei LMS. We already did that many times in the past, and even in this project we're doing the same thing again already, see: 1803-gh-Automattic/sensei-pro - We had some PRs similar to that in the past projects we worked on, like in the Interactive Video Blocks, for instance.

Let me start saying that I know what is the usual approach we have followed previously (using filters) but the reason to no use that and check the classes instead was because of the corner case described. If, as you said, we are ok with that corner case (users with an outdated sensei pro will see the upgrade cta) I will changed that part and rely on a filter 👍 

> That's why we don't need anything like a Sensei_Pro_Detector, at all. In practice, the less the code knows about Sensei Pro, the better 

Despite the above I am still not convinced about the "we don't need anything like `Sensei_Pro_Detector`.
I could add yet another filter (in fact I would need two filters: one for this CTA and another one for #5744 ) but I think it might be too specific.
I will try to explain myself: I agree on "the less that Sensei Core knows about Sensei Pro the better" but in these specific situations both are for a "Sensei Pro" promo and an upgrade to "Sensei Pro". They are so strictly linked to Sensei Pro that I am tempted to have a filter like `sensei_is_sensei_pro_active`.
In that case I would keep the `Sensei_Pro_Detector` but change the logic to be based on a filter instead of the class existence check.
WDYT?

> especially because any test focusing on detecting that easily breaks in a dev environment like the one I have thanks to the fact that Sensei Pro and Sensei Core is both always installed, and, well, I like running tests in my own machine

I don't understand what tests are you referring to here. There shouldn't be any tests breaking based on Sensei Pro being installed or not 🤔  but maybe I am missing something

aaronfc committed:
Replace class check with filter.

aaronfc commented:
@fjorgemota went ahead and implemented it with a filter as I described in my last comment: 71e51c14a

Don't feel forced to accept this solution, I am open to keep discussing it :)

fjorgemota commented on includes/admin/home/help/class-sensei-home-help-item.php:
```diff
@@ -0,0 +1,95 @@
+	 * @param string|null                      $icon       The icon. But default null and will use the default icon (if any).
+	 * @param Sensei_Home_Help_Extra_Link|null $extra_link An optional link with extra information.
+	 */
+	public function __construct( string $title, $url, string $icon = null, Sensei_Home_Help_Extra_Link $extra_link = null ) {
```
Some improvements to the typing system here:

```suggestion
	public function __construct( string $title, ?string $url, ?string $icon = null, ?Sensei_Home_Help_Extra_Link $extra_link = null ) {
```

aaronfc replied:
Done! 1ab40703e

I forgot we had PHP 7.2 as minimum 🎉

fjorgemota reviewed (commented):
Hey there.

> Let me start saying that I know what is the usual approach we have followed previously (using filters) but the reason to no use that and check the classes instead was because of the corner case described. If, as you said, we are ok with that corner case (users with an outdated sensei pro will see the upgrade cta) I will changed that part and rely on a filter +1

Fair enough, let's use a filter instead because supporting older sensei pro versions will not be needed. :) 

> I will try to explain myself: I agree on "the less that Sensei Core knows about Sensei Pro the better" but in these specific situations both are for a "Sensei Pro" promo and an upgrade to "Sensei Pro". They are so strictly linked to Sensei Pro that I am tempted to have a filter like sensei_is_sensei_pro_active.
> In that case I would keep the Sensei_Pro_Detector but change the logic to be based on a filter instead of the class existence check.
> WDYT?

That's definitely a possibility.

Right now, we have some code like that in Sensei Pro already, see the file `assets/upsell-hooks.js` in Sensei Pro. :) 

Another possibility is something similar to what I suggested in https://github.com/Automattic/sensei/pull/5792#pullrequestreview-1122846548 - using filters to filter on the content/output instead to avoid logic directly to Sensei Pro here.

But your approach is fine, too. :+1: 

> I don't understand what tests are you referring to here. There shouldn't be any tests breaking based on Sensei Pro being installed or not thinking but maybe I am missing something

I thought `class_exists` would immediately fail while running tests using a WP installation with both Sensei and Sensei Pro installed, but it doesn't. So...my fail, here, hehe.

--- 

I took an extra review, but I plan on reviewing more tomorrow. Just wanted to reply here as soon as possible. :)

fjorgemota commented:
I added @jom and @renatho so they can also take an extra look on this PR. Personally, I don't have a lot of experience working with designing backend APIs in WordPress, so their experience is welcome. :)

aaronfc committed:
Update with `feature/new-onboarding-experience`.

aaronfc committed:
Remove `Sensei_Pro_Detector` and add `sensei_home_support_ticket_creation_upsell_show` filter.

- Filter `sensei_home_support_ticket_creation_upsell_show` has been
  added to allow for upsell action in Sensei Home to be deactivated.

aaronfc commented:
Hey, I talked last night with my pillow and it convinced me that filters are a better approach.
I have made the changes in 0737c03ae and updated the opening message to include a reference to the new filter and updated instructions on how to test.

Thanks for raising this up, @fjorgemota !

aaronfc committed:
Update types to include Nullables.

aaronfc committed:
Fix access level for tearDown method.

fjorgemota reviewed (approved):
LGTM

renatho commented:
> I added `@jom` and `@renatho` so they can also take an extra look on this PR. Personally, I don't have a lot of experience working with designing backend APIs in WordPress, so their experience is welcome. :)

@fjorgemota, I think you mean the `Sensei_REST_API_Home_Controller`, right? If yes, it looks good to me! :)

fjorgemota commented:
@renatho Hey, yeah. I already approved it. Thanks for looking into it, though! :D
